packages/material-ui/src/FormControl/FormControl.js

```diff
@@ -111,119 +111,34 @@ const FormControl = React.forwardRef(function FormControl(inProps, ref) {
-
-    if (children) {
-      React.Children.forEach(children, (child) => {
-        if (!isMuiElement(child, ['Input', 'Select'])) {
-          return;
-        }
-
-        const input = isMuiElement(child, ['Select']) ? child.props.input : child;
-
-        if (input && isAdornedStart(input.props)) {
-          initialAdornedStart = true;
-        }
-      });
-    }
-    return initialAdornedStart;
-  });
-
-  const [filled, setFilled] = React.useState(() => {
-    // We need to iterate through the children and find the Input in order
-    // to fully support server-side rendering.
-    let initialFilled = false;
-
-    if (children) {
-      React.Children.forEach(children, (child) => {
-        if (!isMuiElement(child, ['Input', 'Select'])) {
-          return;
-        }
-
-        if (isFilled(child.props, true)) {
-          initialFilled = true;
-        }
-      });
-    }
-
-    return initialFilled;
-  });
-
-  const [focusedState, setFocused] = React.useState(false);
-  if (disabled && focusedState) {
-    setFocused(false);
-  }
-
-  const focused = visuallyFocused !== undefined && !disabled ? visuallyFocused : focusedState;
-
-  let registerEffect;
-  if (process.env.NODE_ENV !== 'production') {
-    // eslint-disable-next-line react-hooks/rules-of-hooks
-    const registeredInput = React.useRef(false);
-    registerEffect = () => {
-      if (registeredInput.current) {
-        console.error(
-          [
-            'Material-UI: There are multiple `InputBase` components inside a FormControl.',
-            'This creates visual inconsistencies, only use one `InputBase`.',
-          ].join('\n'),
-        );
-      }
-
-      registeredInput.current = true;
-      return () => {
-        registeredInput.current = false;
-      };
-    };
-  }
-
-  const onFilled = React.useCallback(() => {
-    setFilled(true);
-  }, []);
-
-  const onEmpty = React.useCallback(() => {
-    setFilled(false);
-  }, []);
-
   const childContext = {
-    adornedStart,
-    setAdornedStart,
     color,
-    disabled,
-    error,
-    filled,
-    focused,
     fullWidth,
-    hiddenLabel,
     size,
-    onBlur: () => {
-      setFocused(false);
-    },
-    onEmpty,
-    onFilled,
-    onFocus: () => {
-      setFocused(true);
-    },
-    registerEffect,
-    required,
     variant,
   };
 
+  const componentsProps = {
+    root: { as: component },
+  };
+
   return (
-    <FormControlContext.Provider value={childContext}>
-      <FormControlRoot
-        as={component}
-        styleProps={styleProps}
-        className={clsx(classes.root, className)}
-        ref={ref}
-        {...other}
-      >
-        {children}
-      </FormControlRoot>
-    </FormControlContext.Provider>
+    <FormControlUnstyled
+      className={clsx(classes.root, className)}
+      component={FormControlRoot}
+      componentsProps={componentsProps}
+      disabled={disabled}
+      error={error}
+      extraContextProperties={childContext}
```

Comment:
Seeing this, it makes me wonder if MD shouldn't be built on the hook API instead of the unstyled one. I'm asking because using the hook directly removes an indirection:

1. it might improve performance as we skip a React layer (we save the ...spread, React element cost overhead). 
2. it makes it easier for developers to follow our sources 

Anyway, I'm not saying we should do it, but it could be a valid option.

Reply:
This was my first idea when developing the Switch. It's actually easier to implement the MD Switch using the hooks. Plus, we lose one layer of abstraction.
The reason I used `SwitchUnstyled` instead of `useSwitch` in https://github.com/mui-org/material-ui/pull/27182 is to have higher test coverage. Since `SwitchUnstyled` uses the hook, by testing `SwitchUnstyled` we also test the hook. I also have the hook implementation ready, so once we're good to merge, we can decide which one to take. 

With FormControl it's a bit different. The existing hook doesn't encapsulate the logic and it's just a context accessor, so here it makes more sense to implement `FormControl` using `FormControlUnstyled`.